### PR TITLE
Refactoring config to use local variables

### PR DESCRIPTION
**Pull request by vishnuchalla**
## Type of change

- [x] Refactor
- [ ] New feature
- [ ] Bug fix
- [ ] CVE fix
- [x] Optimization
- [ ] Documentation Update
- [ ] Configuration Update
- [ ] Bump-up dependent library

## Related Tickets & Documents

- Related Issue # https://issues.redhat.com/browse/OLS-379

## Checklist before requesting a review

- [x] I have performed a self-review of my code.
- [x] PR has passed all pre-merge test jobs.
- [x] If it is a core feature, I have added thorough tests.

## Testing
Tested integrating with operator. Works as expected.
```
[vchalla@vchalla-thinkpadp1gen2 ~]$ oc logs pod/lightspeed-app-server-5c8646c79f-d994z -f
Defaulted container "lightspeed-service-api" out of: lightspeed-service-api, lightspeed-service-user-data-collector
There was a problem when trying to write in your cache folder (/.cache/huggingface/hub). You should set the environment variable TRANSFORMERS_CACHE to a writable directory.
2024-05-03 06:35:16,885 [__main__:runner.py:49] INFO: Config loaded from /etc/ols/..2024_05_03_06_33_24.3598472738/olsconfig.yaml
2024-05-03 06:35:16,886 [sentence_transformers.SentenceTransformer:SentenceTransformer.py:113] INFO: Load pretrained SentenceTransformer: /app-root/embeddings_model
2024-05-03 06:35:19,879 [sentence_transformers.SentenceTransformer:SentenceTransformer.py:231] INFO: 2 prompts are loaded, with the keys: ['query', 'text']
2024-05-03 06:35:19,880 [ols.src.rag_index.index_loader:index_loader.py:65] INFO: Setting up settings for index load...
LLM is explicitly disabled. Using MockLLM.
2024-05-03 06:35:22,789 [ols.src.rag_index.index_loader:index_loader.py:68] INFO: Setting up storage context for index load...
2024-05-03 06:35:22,792 [faiss.loader:loader.py:74] INFO: Loading faiss with AVX512 support.
2024-05-03 06:35:23,387 [faiss.loader:loader.py:76] INFO: Successfully loaded faiss with AVX512 support.
2024-05-03 06:35:23,402 [root:base.py:99] INFO: Loading llama_index.vector_stores.faiss.base from /app-root/vector_db/ocp_product_docs/4.15/default__vector_store.json.
2024-05-03 06:35:23,582 [ols.src.rag_index.index_loader:index_loader.py:81] INFO: Loading vector index...
2024-05-03 06:35:23,582 [llama_index.core.indices.loading:loading.py:66] INFO: Loading indices with ids: ['ocp-product-docs-4_15']
2024-05-03 06:35:24,486 [ols.src.rag_index.index_loader:index_loader.py:86] INFO: Vector index is loaded.
2024-05-03 06:35:34,575 [ols.app.main:main.py:30] INFO: Embedded Gradio UI is disabled. To enable set `enable_dev_ui: true` in the `dev_config` section of the configuration file.
INFO:     Started server process [1]
INFO:     Waiting for application startup.
INFO:     Application startup complete.
INFO:     Uvicorn running on http://0.0.0.0:8080 (Press CTRL+C to quit)
2024-05-03 06:35:55,608 [ols.utils.auth_dependency:auth_dependency.py:192] WARNING: Auth checks disabled, skipping
2024-05-03 06:35:55,609 [ols.app.endpoints.ols:ols.py:82] INFO: User ID 00000000-0000-0000-0000-000000000000
2024-05-03 06:35:55,609 [ols.app.endpoints.ols:ols.py:147] INFO: b201e7c8-9ea8-42ba-a864-ddf8999c0d0e New conversation
2024-05-03 06:35:55,609 [ols.app.endpoints.ols:ols.py:88] INFO: b201e7c8-9ea8-42ba-a864-ddf8999c0d0e Incoming request: write a deployment yaml for the mongodb image
2024-05-03 06:35:55,609 [ols.src.query_helpers.question_validator:question_validator.py:48] INFO: b201e7c8-9ea8-42ba-a864-ddf8999c0d0e call settings: conversation_id: b201e7c8-9ea8-42ba-a864-ddf8999c0d0e, query: write a deployment yaml for the mongodb image, provider: bam, model: ibm/granite-13b-chat-v2, verbose: False
2024-05-03 06:35:56,220 [httpx:_client.py:1758] INFO: HTTP Request: GET https://bam-api.res.ibm.com/v2/text/generation/limits?version=2023-11-22 "HTTP/1.1 200 OK"
2024-05-03 06:35:58,029 [httpx:_client.py:1758] INFO: HTTP Request: POST https://bam-api.res.ibm.com/v2/text/generation?version=2024-03-19 "HTTP/1.1 200 OK"
2024-05-03 06:35:58,046 [genai.extensions.langchain.llm:llm.py:200] INFO: Output of GENAI call: ALLOWED

Batches: 100%|██████████| 1/1 [00:06<00:00,  6.21s/it]
2024-05-03 06:36:04,666 [httpx:_client.py:1758] INFO: HTTP Request: GET https://bam-api.res.ibm.com/v2/text/generation/limits?version=2023-11-22 "HTTP/1.1 200 OK"
INFO:     Shutting down
INFO:     Waiting for connections to close. (CTRL+C to force quit)
2024-05-03 06:36:24,886 [httpx:_client.py:1758] INFO: HTTP Request: POST https://bam-api.res.ibm.com/v2/text/generation?version=2024-03-19 "HTTP/1.1 200 OK"
2024-05-03 06:36:24,888 [genai.extensions.langchain.llm:llm.py:200] INFO: Output of GENAI call: ```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: mongodb-deployment
spec:
  selector:
    matchLabels:
      app: mongodb
  replicas: 3
  template:
    metadata:
      labels:
        app: mongodb
    spec:
      containers:
      - name: mongodb
        image: mongo:4.4
        ports:
        - containerPort: 27017

This YAML file creates a deployment called `mongodb-deployment` with three replicas of the MongoDB image (version 4.4). The container exposes port 27017, which is the default port for MongoDB. Make sure to replace the image name and port number according to your requirements.

Keep in mind that this is just an example, and you might need to adjust the configuration based on your specific use case, such as adding environment variables, configuring storage, or setting up network policies.

For more information on OpenShift and Kubernetes Deployments, you can refer to the official documentation:

* [OpenShift Deployments](https://docs.openshift.com/container-platform/4.5/dev_guide/deploying_applications/index.html)
* [Kubernetes Deployments](https://kubernetes.io/docs/concepts/workloads/controllers/deployment/)
2024-05-03 06:36:24,888 [ols.app.endpoints.ols:ols.py:229] INFO: b201e7c8-9ea8-42ba-a864-ddf8999c0d0e Storing conversation history.

```

### CODE REVIEW

**Review by onmete — COMMENTED** [review]
For discussion:
I would find something like this
```python
from ols.configuration import config

config.ols_config.user_data_collection.feedback_disabled
config.rag_index
config.ols_config.default_model
config.dev_config.run_on_localhost
...
```
more understandable and straightforward.
No additional shortcuts for config subdivisions (llm_config,dev_config,...) that only complicates the whole thing.
The `config` can be as simple as a module constant (initialized `Config` instance).

Try to imagine the `Config` (base pydantic model) as a wrapper for the individual configurations - like `olsconfig.yaml` for its content.

File: runner.py [hunk 1]
```diff
@@ -19,9 +19,9 @@ def configure_hugging_face_envs(ols_config) -> None:
         and hasattr(ols_config.reference_content, "embeddings_model_path")
         and ols_config.reference_content.embeddings_model_path
     ):
-        os.environ["TRANSFORMERS_CACHE"] = (
-            ols_config.reference_content.embeddings_model_path
-        )
+        os.environ[
+            "TRANSFORMERS_CACHE"
+        ] = ols_config.reference_content.embeddings_model_path
```
Comment (onmete) [inline runner.py:24]: `config_manager.get_ols_config().reference_content.embeddings_model_path`?

File: ols/utils/config.py [hunk 2]
```diff
@@ -4,74 +4,127 @@
...
+
+    def __new__(cls):
+        """Ensure only one instance of ConfigManager exists (Singleton pattern)."""
+        if cls._instance is None:
+            cls._instance = super().__new__(cls)
+        return cls._instance
+
+    def __init__(self):
+        """Initialize instance attributes only if not already initialized."""
+        if not hasattr(self, "config"):  # Check to avoid re-initialization
```
Comment (tisnik) [inline ols/utils/config.py:29]: is it needed when we know in advance, that we have singleton?

File: ols/utils/config.py [hunk 3]
```diff
@@ -4,74 +4,127 @@
...
+
+    def init_vector_index(self) -> None:
+        """Initialize vector index."""
+        # TODO: OLS-380 Config object mirrors configuration
+        self.rag_index = IndexLoader(self.ols_config.reference_content).vector_index
+
+    def get_config(self) -> config_model.Config:
+        """Get the current configuration."""
+        if self.config is None:
+            raise ValueError("Configuration has not been initialized.")
```
Comment (tisnik) [inline ols/utils/config.py:89]: is not it too late? I mean - we can:

1. have proper config object
2 or the app failed already in `runner.py`

File: tests/unit/utils/test_token_handler.py [hunk 4]
```diff
@@ -237,61 +237,68 @@ def test_limit_conversation_history(self):
...
         assert truncated
 
         # try to truncate to zero tokens
-        truncated_history, truncated = (
-            self._token_handler_obj.limit_conversation_history(history, 0)
-        )
+        (
+            truncated_history,
+            truncated,
+        ) = self._token_handler_obj.limit_conversation_history(history, 0)
```
Comment (tisnik) [inline tests/unit/utils/test_token_handler.py:292]: hmm does it pass through Black?

**Comment by tisnik** [comment]
> For discussion: I would find something like this
> 
> ```python
> from ols.configuration import config
> 
> config.ols_config.user_data_collection.feedback_disabled
> config.rag_index
> config.ols_config.default_model
> config.dev_config.run_on_localhost
> ...
> ```
> 
> more understandable and straightforward. No additional shortcuts for config subdivisions (llm_config,dev_config,...) that only complicates the whole thing. The `config` can be as simple as a module constant (initialized `Config` instance).
> 
> Try to imagine the `Config` (base pydantic model) as a wrapper for the individual configurations - like `olsconfig.yaml` for its content.

I agree with @onmete

File: ols/utils/config.py (hunk 3 above)
Comment (tisnik) [reply ols/utils/config.py:89]: i.e. there should not be intermediate state, when the config is half-initialized

File: ols/app/endpoints/ols.py [hunk 5]
```diff
@@ -109,7 +110,7 @@ def conversation_request(
 
     store_conversation_history(user_id, conversation_id, llm_request, response)
 
-    if config.ols_config.user_data_collection.transcripts_disabled:
+    if config_manager.get_ols_config().user_data_collection.transcripts_disabled:
```
Comment (tisnik) [inline ols/app/endpoints/ols.py:113]: if we use Pydantic approach (w/o getters), most of the code won't need to change and this PR might be more minimalistic

**Comment by vishnuchalla** [comment]
@onmete @tisnik Please bear with me as this PR is still in progress. I will retag once I update it. Thanks

**Comment by vishnuchalla** [comment]
> For discussion: I would find something like this
> 
> ```python
> from ols.configuration import config
> 
> config.ols_config.user_data_collection.feedback_disabled
> config.rag_index
> config.ols_config.default_model
> config.dev_config.run_on_localhost
> ...
> ```
> 
> more understandable and straightforward. No additional shortcuts for config subdivisions (llm_config,dev_config,...) that only complicates the whole thing. The `config` can be as simple as a module constant (initialized `Config` instance).
> 
> Try to imagine the `Config` (base pydantic model) as a wrapper for the individual configurations - like `olsconfig.yaml` for its content.

Please help me understand on what you exactly meant by `pyndantic` approach here, how will we even be able to invoke config.ols_config without `ols_config` being a globally available attribute of the config class?

**Comment by onmete** [comment]
It means using the wrapper class `ols.app.models.config.Config` we already have, as the config entrypoint. You can access every part of the configuration from there (if something is missing, it should be added).

Regarding initialization, something like this could work.
```python
# in the config.py
with open(os.environ.get("OLS_CONFIG_FILE", "olsconfig.yaml"), "r", encoding="utf-8") as f:
    data = yaml.safe_load(f)
config = config_model.Config(data)
```
The `config` is module constant/singleton.
It is just a though, there might be a better way how to set it up.

**Comment by vishnuchalla** [comment]
Further changes to be done using pydantic approach. poc pr: https://github.com/openshift/lightspeed-service/pull/877

**Pull request closed without merging**